// The file defines the protos for specificying the evaluation of BenchPress.
syntax = "proto2";

package clgen;

option go_package = "clgenpb";
option java_multiple_files = true;
option java_outer_classname = "EvaluatorProto";
option java_package = "com.clgen";

// import "deeplearning/clgen/proto/model.proto";
// import "deeplearning/clgen/proto/corpus.proto";

// Specification of app's evaluation pipeline.
message Evaluation {
  // Define a list of different evaluators to run.
  repeated evaluators evaluator = 1;
}

// Instance of a single evaluation pass.
message evaluators {
  // Define all different evauators supported.
  oneof evaluator {
    LogFile       log_file        = 1;
    KAverageScore k_average_score = 2;
    MinScore      min_score       = 3;
    AnalyzeTarget analyze_target  = 4;
  }
}

// 1
// Dump log file for all databases provided.
message LogFile {
  optional string          target    = 1;
  repeated DatabasesGroups db_groups = 2;
}

// 2
// Calculate and plot the average score of the top-K best samples from each DB group.
message KAverageScore {
  optional string          target         = 1;
  optional string          feature_space  = 2;
  optional int32           top_k          = 3;
  repeated DatabasesGroups db_groups      = 4;
}

// 3
// Calculate and plot the minimum score from each DB group.
message MinScore {
  optional string          target         = 1;
  optional string          feature_space  = 2;
  repeated DatabasesGroups db_groups      = 3;
}

// 4
// Analyze the size of target bechmraks in number of instructions and token length.
message AnalyzeTarget {
  repeated string target    = 1;
  optional string tokenizer = 2;
}

// 5
// Map on 2-Dimensional space the number of memory vs computational instructions from Grewe's features.
message CompMemGrewe {
  optional string          target    = 1;
  repeated DatabasesGroups db_groups = 2;
}

// Define a group of Databases that will be considered jointly.
message DatabasesGroups {
  optional string group_name = 1;
  optional string db_type    = 2;
  repeated string databases  = 3;
}
